arch/arm/arm64/head.S

```diff
@@ -85,6 +104,170 @@ ENTRY(_start)
     load_paddr  x0, __start_bss
     bl      flush_dcache_range
 
-    /* Start an infinite loop */
-    PRINT("- Infinite loop -\n")
-2:  b   2b
+    PRINT("- Setup stack -\n")
+    ldr     x1, =stack_end
+    mov     sp, x1
+
+    /* Save boot arguments */
+    ldr     x0, =boot_data
+    stp     x21, x20, [x0]
+
+    PRINT("- Jump to C world-\n")
+    b       xtf_main
+ENDFUNC(_start)
+
+ENTRY(_mmu_disable)
+    dsb     sy
+
+    /* Turn off D-cache and MMU */
+    mrs     x2, sctlr_el1
+    bic     x2, x2, #SCTLR_M
+    bic     x2, x2, #SCTLR_C
+    msr     sctlr_el1, x2
+    isb
+
+    ret
+ENDFUNC(_mmu_disable)
+
+ENTRY(_cpu_setup)
+    dsb     sy
+
+    /* Set up memory attribute type tables */
+    ldr     x0, =MAIRVAL
+    msr     mair_el1, x0
+    isb
+
+    ret
+ENDFUNC(_cpu_setup)
+
+/* Save state */
+.macro entry_trap
+    sub     sp, sp, #(272 - 240)     /* offset: spsr_el1 - lr */
+    stp     x28, x29, [sp, #-16]!
+    stp     x26, x27, [sp, #-16]!
+    stp     x24, x25, [sp, #-16]!
+    stp     x22, x23, [sp, #-16]!
+    stp     x20, x21, [sp, #-16]!
+    stp     x18, x19, [sp, #-16]!
+    stp     x16, x17, [sp, #-16]!
+    stp     x14, x15, [sp, #-16]!
+    stp     x12, x13, [sp, #-16]!
+    stp     x10, x11, [sp, #-16]!
+    stp     x8, x9, [sp, #-16]!
+    stp     x6, x7, [sp, #-16]!
+    stp     x4, x5, [sp, #-16]!
+    stp     x2, x3, [sp, #-16]!
+    stp     x0, x1, [sp, #-16]!
+
+    add     x21, sp, #272           /* offset: spsr_el1 */
+    stp     lr, x21, [sp, #240]     /* offset: lr */
+    mrs     x21, elr_el1
+    mrs     x22, spsr_el1
+    stp     x21, x22, [sp, #256]    /* offset: pc */
+.endm
+
+/* Restore state */
+.macro exit_trap
+    ldp     x21, x22, [sp, #256]    /* offset: pc */
+    ldp     x0, x1, [sp], #16
+    ldp     x2, x3, [sp], #16
+    ldp     x4, x5, [sp], #16
+    ldp     x6, x7, [sp], #16
+    ldp     x8, x9, [sp], #16
+
+    msr     elr_el1, x21            /* set up the return data */
+    msr     spsr_el1, x22
+
+    ldp     x10, x11, [sp], #16
+    ldp     x12, x13, [sp], #16
+    ldp     x14, x15, [sp], #16
+    ldp     x16, x17, [sp], #16
+    ldp     x18, x19, [sp], #16
+    ldp     x20, x21, [sp], #16
+    ldp     x22, x23, [sp], #16
+    ldp     x24, x25, [sp], #16
+    ldp     x26, x27, [sp], #16
+    ldp     x28, x29, [sp], #16
+
+    ldr     lr, [sp], #32           /* offset: spsr_el1 - lr */
+    eret
+.endm
+
+/* Bad abort numbers */
+#define BAD_SYNC    0
+#define BAD_IRQ     1
+#define BAD_FIQ     2
+#define BAD_ERROR   3
+
+.macro invalid, reason
```

Comment:
Some of the code in this commit (this is one example) looks suspiciously similar to Xen code. Did you look at Xen code while writing XTF?

Reply:
I looked at Xen and Linux. If there is anything wrong with it, please tell me.

Reply:
> For some parts, I looked at Xen and Linux. If there is anything wrong with it, please tell me.

There is a licensing issue. AFAIK, it is fine to import BSD code in GPL but the invert is not. So when writing BSD code, you should avoid to read Linux/Xen code to avoid been biased.

If you look for OS example, then this should come from a BSD OS.